components/search/src/environment_matching.rs

```diff
@@ -816,29 +652,359 @@ mod tests {
+        );
+
+        assert!(
+            matches_user_environment(
+                &crate::JSONVariantEnvironment {
+                    channels: vec![SearchUpdateChannel::Nightly, SearchUpdateChannel::Release],
+                    ..Default::default()
+                },
+                &SearchUserEnvironment {
+                    update_channel: SearchUpdateChannel::Release,
+                    ..Default::default()
+                }
+            ),
+            "Should return true when the channel matches one in the environment when there are multiple"
+        );
+
+        assert!(
+            !matches_user_environment(
+                &crate::JSONVariantEnvironment {
+                    channels: vec![SearchUpdateChannel::Nightly],
+                    ..Default::default()
                 },
                 &SearchUserEnvironment {
                     locale: "fi".into(),
                     region: "fr".into(),
                     update_channel: SearchUpdateChannel::Default,
-                    distribution_id: String::new(),
+                    distribution_id: "distro-4".into(),
                     experiment: String::new(),
                     app_name: SearchApplicationName::Firefox,
                     version: String::new(),
+                    device_type: SearchDeviceType::None,
                 }
             ),
-            "Should return true when the distribution and region matches the environment"
+            "Should return false when the channel does not match any in the environment"
+        );
+
+        assert!(
+            matches_user_environment(
+                &crate::JSONVariantEnvironment {
+                    regions: vec!["fr".to_string()],
+                    channels: vec![SearchUpdateChannel::Default],
+                    ..Default::default()
+                },
+                &SearchUserEnvironment {
+                    locale: "fi".into(),
+                    region: "fr".into(),
+                    update_channel: SearchUpdateChannel::Default,
+                    ..Default::default()
+                }
+            ),
+            "Should return true when the channel and region matches the environment"
+        );
+    }
+
+    #[test]
+    fn test_matches_user_environment_experiment() {
+        assert!(
+            matches_user_environment(
+                &crate::JSONVariantEnvironment {
+                    experiment: "warp-drive".to_string(),
+                    ..Default::default()
+                },
+                &SearchUserEnvironment {
+                    experiment: "warp-drive".to_string(),
+                    ..Default::default()
+                }
+            ),
+            "Should return true when the experiment matches the one in the environment"
+        );
+
+        assert!(
+            !matches_user_environment(
+                &crate::JSONVariantEnvironment {
+                    experiment: "warp-drive".to_string(),
```

Comment:
TIL - Warp drive was a technology that allowed space travel at faster-than-light speeds. 😄